### PR TITLE
Split change outputs with asset quantities exceeding the maximum

### LINKED ISSUE #2532: Major Wallet locking Issue with NativeAssets overflow in UTXOs
## Essential Information

Daedalus 3.3.0#16328 Frontend
Public-Testnet

# Steps to Reproduce

Wallet-Owner:
* Create a new Wallet in Daedalus/Cardano-Wallet via SeedPhraseMethod
* Sends 10M Ada to fund the Wallet to the first Receive-Address

Now a bad actor starts to come in:
* Sends 18446744073709551614 XCoins to the first Receive-Address
* Sends again 18446744073709551614 XCoins to the first Receive-Address
* Sends again 18446744073709551614 XCoins to the first Receive-Address

Now the wallet looks like this:
image

Wallet-Owner:
* Sees the 6 additional ADA but yes, they are just there, Wallet-Owner is happy about free ADA
* Wallet-Owner wants to send 8M ADA out of the Wallet
* Transaction fails

image

Why does it fail? Because cardano-wallet is also using two of the additional UTXOs that the bad actor sent over. It tries to consolidate the Native Assets of course, but the sum of two of the UTXOs generates an overflow. Because the maximum possible value is 18446744073709551614 or 18446744073709551615, doesn't matter.

The Wallet is now locked up. 

You can maybe send out smaller pieces with the hope that the auto UTXO picking algo is not using any of the asset UTXOs again.

You also can't get rid of the assets, because first - Daedalus does not support it right now, but second - You will also end up with more than the maximum amount in the transaction and the node will fail.

# Resolution

We need a much better UTXO selection algo ASAP! Or even an advanced mode in the wallets to hand-pick the UTXOs you wanna use. If this happens on a cli based account like for SPOs, this is solvable. But a normal User can't solve this. Funds are locked up, or kind of. But imagin the reaction of the Wallet-Owner. Also, restoring the wallet on another Frontend like Yoroi or Adalite results in the same error.

I can provide the DevTeam a SeedPhrase if someone wants to restore the TestWallet i created.

# Jira Issue number

ADP-726, ADP-826

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

https://github.com/input-output-hk/cardano-wallet/issues/2532
ADP-726

# Overview

Although quantities of individual assets are effectively unlimited, a transaction on the blockchain can never include an asset quantity greater than `maxBound :: Word64`.

This PR tweaks the coin selection algorithm to detect change bundles containing excessively large asset quantities.

If such a change bundle is detected, we now split the change bundle up into smaller bundles using [equipartitioning](https://en.wiktionary.org/wiki/equipartition).

# Example

Let's suppose the maximum allowable token quantity is **`5`**, and we have a change map with the following quantities:

```haskell
[("a", 11), ("b", 5), ("c", 12)]
```

In this case, we must divide the map into **_at least three_** smaller maps in order to not exceed the maximum allowable token quantity.

Under the equipartitioning scheme, this would give us:

```haskell
[("a", 3), ("b", 1), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
```

Note that while the overall sum is preserved, the individual bundles are almost equal, **_but not quite_**: this is because `11` and `5` are not divisible by `3`. We must therefore accept a small loss of proportionality in the result.

# Details

An **_equipartition_** of a bundle **_b_** is a _partition_ into multiple bundles, where for every asset **_a_** in the set of assets contained in **_b_**, the difference between the following quantities is either _zero_ or _one_ :

- The smallest quantity of asset **_a_** in the resultant bundles
- The greatest quantity of asset **_a_** in the resultant bundles

In order to determine the number of parts in which to split a given bundle, we choose the **_smallest_** number of parts that still allows us satisfy the goal of not exceeding the maximum allowable quantity in any given bundle.

# Performance

In order to avoid evaluating a partition for every single change output, we **_short circuit_** in the event that there is no token quantity greater than the maximum allowable quantity:

```haskell
equipartitionTokenMapWithMaxQuantity m (TokenQuantity maxQuantity)
    | maxQuantity == 0 =
        maxQuantityZeroError
    | currentMaxQuantity <= maxQuantity =
        m :| []
    | otherwise =
        equipartitionTokenMap m (() :| replicate extraPartCount ())
```

# Testing

## Property tests

Equipartitioning behaviour is tested by the following property tests:
- `prop_equipartitionTokenBundle*`
- `prop_equipartitionTokenMap*`

## Unit tests

As a sanity check, this PR also provides unit tests for `performSelection` with inputs containing token quantities that are close to the maximum, and demonstrates that change bundles are correctly partitioned in the results.

### CODE REVIEW

File: .hlint.yaml [hunk 1]
```diff
@@ -54,6 +54,7 @@
 - ignore: {name: "Redundant multi-way if"} # Using this style might be appropriate given the circumstances.
 - ignore: {name: "Use const"} # Sometimes using a lambda expression is simpler and clearer.
 - ignore: {name: "Redundant id"} # Sometimes this is useful for vertical alignment of elements.
+- ignore: {name: "Use camelCase"} # Sometimes this is useful in test code.
```
Comment (rvl) [inline .hlint.yaml]: I think we normally want this hlint enabled for application code.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 2]
```diff
@@ -1219,6 +1436,16 @@ newtype AssetCount a = AssetCount
...
 
+--------------------------------------------------------------------------------
+-- Constants
+--------------------------------------------------------------------------------
+
+-- | The greatest token quantity that can be encoded within an output bundle of
+--   a transaction.
+--
+maxTokenQuantity :: TokenQuantity
+maxTokenQuantity = TokenQuantity $ fromIntegral (maxBound :: Word64)
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs]: It's a bit of a mouthful, but renaming maxTokenQuantity -> maxTokenQuantityInTxOut will prevent misleading any readers.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 3]
```diff
@@ -855,11 +870,25 @@ makeChange minCoinFor requiredCost mExtraCoinSource inputBundles outputBundles
...
         -- asset count:
         & NE.zipWith (\m1 (m2, c) -> (m1 <> m2, c))
             changeForNonUserSpecifiedAssets
+        -- Finally, if there are any maps with excessive token quantities, then
+        -- split these maps up along with their corresponding output coins:
+        & splitMapsWithExcessiveQuantities
+      where
+        splitMapsWithExcessiveQuantities
+            :: NonEmpty (TokenMap, Coin) -> NonEmpty (TokenMap, Coin)
+        splitMapsWithExcessiveQuantities =
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:884]: I would prefer if the unit testing were based on properties of this function.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 4]
```diff
@@ -1145,6 +1174,194 @@ makeChangeForCoin targets excess =
...
+--
+equipartitionCoin
+    :: HasCallStack
+    => Coin
+    -- ^ The coin to be partitioned.
+    -> NonEmpty a
+    -- ^ Represents the number of portions in which to partition the coin.
+    -> NonEmpty Coin
+    -- ^ The partitioned coins.
+equipartitionCoin c =
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:1213]: Given the max supply of ada, there's no need to apply this fix to Coin.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 5]
```diff
@@ -1145,6 +1174,194 @@ makeChangeForCoin targets excess =
...
+--
+equipartitionTokenBundle
+    :: HasCallStack
+    => TokenBundle
+    -- ^ The bundle to be partitioned.
+    -> NonEmpty a
+    -- ^ Represents the number of portions in which to partition the bundle.
+    -> NonEmpty TokenBundle
+    -- ^ The partitioned bundle.
+equipartitionTokenBundle (TokenBundle c m) count =
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs]: This function shouldn't be needed, right? Only equipartitionTokenMap.

**Review by rvl — COMMENTED** [review]
This will work but it was confusing to review and seems like a lot of code for something which is basically this:
```haskell
example :: [(String, Natural)]
example = [("a", 11), ("b", 5), ("c", 12)]

-- >>> limitSize 5 example
-- [[("a",5),("b",5),("c",5)],[("a",5),("c",5)],[("a",1),("c",2)]]
limitSize :: (Traversable t, Integral n) => n -> [t n] -> [[t n]]
limitSize lim = group . map (fmap (splitNum lim))
  where
    group = L.transpose . map sequence

splitNum :: Integral n => n -> n -> [n]
splitNum q a = replicate (fromIntegral n) q ++ if r == 0 then [] else [r]
  where
    n = a `div` q
    r = a `mod` q
```
So is it possible to adjust this limitSize slightly so that it can be applied to TokenMap?

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:1213]: > Given the max supply of ada, there's no need to apply this fix to Coin.

The point here is that we need to divide output coins **_together_** with change maps, so that we can retain proportionality. (Each change map is paired with an output coin, so that when we assign coins to change maps, we can assign them in proportion to the sizes of the coins in the original outputs.)

Let's say we need to partition the following token map into two:

- TokenMap A:
    - Banana: 100
    - Apple: 10

This would give us two token maps with the following tokens:

- TokenMap A1:
    - Banana: 50
    - Apple: 5
- TokenMap A2:
    - Banana: 50
    - Apple: 5

However, the change generation algorithm pairs each change token map with a coin from the original outputs, so that when we eventually assign ada values to change maps (to make change bundles), we can assign them in proportion to the sizes of the ada in the original outputs.

So we also need to divide the coin value.

Let's say this particular change map was paired with an output coin value of `100`. That will mean that each divided change map will be paired with a coin value of `50`.

For the sake of convenience, when dividing the pair (change map, output coin value), we treat it as a token bundle, so that we can use the `equipartitionTokenBundleWithMaxQuantity` function.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs (hunk 5 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs]: > This function shouldn't be needed, right? Only equipartitionTokenMap.

You're right -- I've deleted it.

Fixed in 607ec07474ba813e9e7adc344c58c28b7f8f8fac

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs]: I agree this is a good idea. 👍🏻 

Fixed in 0cfb86f7ffc555d740944bbac506e2676a73f701

**Comment by jonathanknowles** [comment]
@rvl wrote:
> So is it possible to adjust this limitSize slightly so that it can be applied to TokenMap?
> ```haskell
> -- >>> limitSize 5 example
> -- [[("a",5),("b",5),("c",5)],[("a",5),("c",5)],[("a",1),("c",2)]]
> limitSize :: (Traversable t, Integral n) => n -> [t n] -> [[t n]]
> limitSize lim = group . map (fmap (splitNum lim))
>   where
>     group = L.transpose . map sequence
> ```

I deliberately avoided the use of `L.transpose` here, as from past experience in using this function, it often gives poor performance in comparison to a single strict fold with an appropriately-written accumulator function.

This is what we've done in `equipartitionTokenMap`:
```hs
equipartitionTokenMap m count =
    F.foldl' accumulate (TokenMap.empty <$ count) (TokenMap.toFlatList m)
  where
    accumulate
        :: NonEmpty TokenMap
        -> (AssetId, TokenQuantity)
        -> NonEmpty TokenMap
    accumulate maps (asset, quantity) = NE.zipWith (<>) maps $
        TokenMap.singleton asset <$>
            equipartitionTokenQuantity quantity count
```
This is called with the appropriate value of `count` by `equipartitionTokenMapWithMaxQuantity`.

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 6]
```diff
@@ -1576,6 +1838,194 @@ unit_makeChangeForUserSpecifiedAsset =
...
+-- Test that natural numbers are equipartitioned fairly:
+--
+-- Each portion must be within unity of the ideal portion.
+--
+prop_equipartitionNatural_fair
+    :: Natural -> NonEmpty () -> Property
+prop_equipartitionNatural_fair n count =
+    prop_equipartitionNatural_fair_inner $ equipartitionNatural n count
+
+prop_equipartitionNatural_fair_inner :: NonEmpty Natural -> Property
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs]: This function can actually now be inlined into `prop_equipartitionNatural_fair`.

File: .hlint.yaml (hunk 1 above)
Comment (jonathanknowles) [reply .hlint.yaml]: Fair enough!

I've moved the suppression of this warning to just the `RoundRobinSpec` module.

Fixed in: b9454787a8184e0e83338602147614f56763c5bd

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs (hunk 3 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:884]: > I would prefer if the unit testing were based on properties of this function.

Fair enough.

Since the critical part of this function relates to list-processing, I've extracted this out as a separate function, for which I've added some simple property tests.

Now the `splitMapsWithExcessiveQuantities` function is simply a wrapper function that bundles and unbundles the `(TokenMap, Coin)` pair.

See 9cf72cba44204032a1238c40e4c7ee156ad6914d.

**Comment by jonathanknowles** [comment]
@rvl wrote:
> ```haskell
> example :: [(String, Natural)]
> example = [("a", 11), ("b", 5), ("c", 12)]
>
> -- >>> limitSize 5 example
> -- [[("a",5),("b",5),("c",5)],[("a",5),("c",5)],[("a",1),("c",2)]]
> limitSize :: (Traversable t, Integral n) => n -> [t n] -> [[t n]]
> limitSize lim = group . map (fmap (splitNum lim))
>   where
>     group = L.transpose . map sequence
> ```

Having looked at this in more detail: your suggestion is different from what this PR is doing, which is to retain the proportionality between quantities of different tokens in a bundle to be partitioned.

So to take your `example` bundle (above), the `equipartitionTokenBundle` function would split it into the following bundles:
```haskell
[("a", 3), ("b", 1), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
```
Note that the original proportions are maintained (modulo rounding).

**Comment by rvl** [comment]
OK thanks for the answers. I am trying to find out why it needs to be much more complicated than the example code I showed. In that example code, the `splitNum` function could/should be modified to divide the value evenly, and List.transpose can easily be swapped for something efficient. What is this proportionality and why is it needed? I thought that "equipartition" meant that the weights are [1,1,1,..], which is even not proportional.

**Comment by jonathanknowles** [comment]
@rvl wrote:
> OK thanks for the answers.

You're welcome.

> I am trying to find out why it needs to be much more complicated than the example code I showed. In that example code, the `splitNum` function could/should be modified to divide the value evenly, and List.transpose can easily be swapped for something efficient.

You're right. Though in doing so, I'm guessing we would still end up with something that is a special case of the already-existing `partitionNatural` function, which is already well-tested, and has all the properties we need, including:
- preservation of the sum
- exact proportionality where possible, but with rounding in the case where values are not evenly divisible.

Reusing this pre-existing function means we can worry a bit less about checking that rounding is correct.

> What is this proportionality and why is it needed?

## Short answer

This was the solution proposed in the [original discussion thread](https://input-output-rnd.slack.com/archives/GBT05825V/p1614051938013600?thread_ts=1614000516.005800&cid=GBT05825V) for this PR. However, since Slack threads are bad for recording design decisions, I've copied and pasted the essential part of that discussion into https://jira.iohk.io/browse/ADP-726. (My apologies for copying it over a bit late.)

## Longer answer

We're trying, to the greatest extent possible, to retain the ratios between asset quantities in the original outputs. If a user specifies an output of `[("assetA", 500), ("assetB", 1000)]` (in a `1:2` ratio), then when generating a change bundle for this output, we try to generate a bundle that has assets in the same `1:2` ratio. Usually, this will also mean that the asset **_quantities_** themselves are also similar to those in the original outputs, as we aim to select around 200% of each asset during the initial selection phase. The idea behind this is that over time, the user's UTxO set distribution will contain entries that are similar to the payments they commonly make.

If we reach the situation where we absolutely have to split up a change bundle to avoid overflow, then there doesn't seem to be a good reason to break this ratio.

Furthermore, retaining the ratio between assets might help to minimize the production of dust quantities.

To give an illustration: if we overflow by just 1, then we'd probably want to avoid partitioning a token quantity into `[maxBound :: Word64, 1]`, because `1` is a value that isn't likely to be very useful when making future coin selections.

@rvl wrote:
> I thought that "equipartition" meant that the weights are `[1,1,1,..]`, which is even not proportional.

You're right, in that the weights are `[1, 1, 1 ..]`: this means we'll generate bundles that are approximately equal to one another.

But proportionality here refers to the fact that every bundle in the partitioned result has a similar ratio of token quantities. For example, if the original bundle has `tokenA:tokenB` in a `1:2` ratio, then the partitioned bundles will also have these tokens in a similar ratio.

So let's take an easy example:

## Example 1

```hs
[("a", 30), ("b", 24), ("c", 18)]
```
The tokens 'a', 'b', and 'c' are in a `5:4:3` ratio.

Suppose the maximum token quantity is `10`. In this case, we'd need to partition the bundle into **_three_** smaller bundles:
```hs
[("a", 10), ("b", 8), ("c", 6)]
[("a", 10), ("b", 8), ("c", 6)]
[("a", 10), ("b", 8), ("c", 6)]
```
The quantities of 'a', 'b', and 'c' in the resulting bundles are still in a `5:4:3` ratio. Since all the numbers are divisible by `3`, the resulting bundles are identical.

## Example 2

Suppose though that the maximum token quantity is `6`. In this case, we'd need to partition into **_five_** smaller bundles:
```hs
[("a", 6), ("b", 4), ("c", 3)]
[("a", 6), ("b", 5), ("c", 3)]
[("a", 6), ("b", 5), ("c", 4)]
[("a", 6), ("b", 5), ("c", 4)]
[("a", 6), ("b", 5), ("c", 4)]
```
Now we hit the situation where `24` and `18` are not evenly divisible by `5`, so we have to round some values down in the result, and some values up. (This is handled automatically by `partitionNatural`.)

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 7]
```diff
@@ -1219,6 +1439,16 @@ newtype AssetCount a = AssetCount
...
 
+--------------------------------------------------------------------------------
+-- Constants
+--------------------------------------------------------------------------------
+
+-- | The greatest token quantity that can be encoded within an output bundle of
+--   a transaction.
+--
+maxTxOutTokenQuantity :: TokenQuantity
+maxTxOutTokenQuantity = TokenQuantity $ fromIntegral (maxBound :: Word64)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:1450]: It's actually half that. The max value that can be encoded must fit in an `int64`, so `9223372036854775807`.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs [hunk 8]
```diff
@@ -1145,6 +1176,195 @@ makeChangeForCoin targets excess =
...
+        maxQuantityZeroError
+    | currentMaxQuantity <= maxQuantity =
+        m :| []
+    | otherwise =
+        equipartitionTokenMap m (() :| replicate extraPartCount ())
+  where
+    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
+
+    extraPartCount :: Int
+    extraPartCount = floor $ pred currentMaxQuantity % maxQuantity
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:1336]: That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above :heavy_check_mark: 

That'd be clearer with a let-binding because this calculation only makes sense with the surrounding context :smirk:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 9]
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
...
+    boundaryTestUTxO =
+      [ (Coin 1_000_000, [(assetA, q1)])
+      , (Coin 1_000_000, [(assetA, q2)])
+      ]
+    boundaryTestInputs =
+      [ (Coin 1_000_000, [(assetA, q1)])
+      , (Coin 1_000_000, [(assetA, q2)])
+      ]
+    boundaryTestChange =
+      [ (Coin 500_000, [(assetA, maxTxOutTokenQuantity)]) ]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:1191]: :heavy_check_mark:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 10]
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
...
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:1220]: :heavy_check_mark:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 11]
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
...
+      , (Coin 1_000_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
+    boundaryTestInputs =
+      [ (Coin 1_000_000, [(assetA, TokenQuantity 1)])
+      , (Coin 1_000_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
+    boundaryTestChange =
+      [ (Coin 250_000, [(assetA, q1)])
+      , (Coin 250_000, [(assetA, q2)])
+      ]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:1252]: :heavy_check_mark:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 12]
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
...
+      , (Coin 1_000_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
+    boundaryTestInputs =
+      [ (Coin 1_000_000, [(assetA, maxTxOutTokenQuantity)])
+      , (Coin 1_000_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
+    boundaryTestChange =
+      [ (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      , (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:1282]: :heavy_check_mark:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs [hunk 13]
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
...
+
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:1283]: It'd have been nice to also show an example where some assets are specified in the outputs, but the resulting change is still going above limit.

**Review by KtorZ — APPROVED** [review]
bors merge

**Comment by KtorZ** [comment]
bors retry

**Comment by gitmachtl** [comment]
I apologize to give you such headache with this issue, i was just testing out the limits. But i think the solution will be worth it. 🙂

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs (hunk 8 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:1336]: > That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above heavy_check_mark

Yes, it does seem odd at first. Though you're absolutely right -- it's this way because of the non-empty list.

The test case for this uses a more straightforward calculation:

```hs
-- | Computes the number of parts that 'equipartitionTokenMapWithMaxQuantity'
--   should return.
--
equipartitionTokenMapWithMaxQuantity_expectedLength
    :: TokenMap -> TokenQuantity -> Int
equipartitionTokenMapWithMaxQuantity_expectedLength
    m (TokenQuantity maxQuantity) =
        max 1 $ ceiling $ currentMaxQuantity % maxQuantity
  where
    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
```

**Comment by jonathanknowles** [comment]
@gitmachtl wrote:
> I apologize to give you such headache with this issue, i was just testing out the limits. But i think the solution will be worth it. slightly_smiling_face

No need to apologise! We are very grateful that you took the time to create a report. It's much better that we can fix this early on, before it inconveniences people attempting to use the wallet.

Thank-you again for your report!
